### DISCUSSION TITLE
How would you deal with large triangles in a mesh?

### DISCUSSION THREAD

**Discussion by stla** [pyvista/pyvista#2540] [Q&A] [answered]
Hello,

 Here is my favorite Hopf torus obtained from a regular 60x65 grid, so there are 60x65x2 triangles:

![HopfRGL](https://user-images.githubusercontent.com/4466543/166198527-acc3709e-f761-4538-96d8-e62b1ebf5d6d.png)

As you can see, the mesh is very dense near the center while the triangles are large at the extremities of the lobes. I obtained this mesh with the R package **rgl**, but I would obtain the same with **pyvista**. However, here is this Hopf torus obtained with **Asymptote**, also using a 60x65 grid:

![HopfTorusMovingColors_texSE](https://user-images.githubusercontent.com/4466543/166198879-0e2548fc-2180-46c4-8c6e-4f2098099aa0.png)

The smoothness is amazing. That's because Asymptote uses splines (Bezier patches or things like that; I don't know this topic yet). So probably it refines the triangles with some splines. 

Are you aware of a Python library which would be helpful for this situation, using something such as Bezier triangles? I found the **bezier** library yesterday but I didn't try it yet. What about implementing spline surfaces in PyVista? There's no such thing in VTK?

**Comment by tkoyama010** [comment, marked as answer]
[`vtkPath`](https://vtk.org/doc/nightly/html/classvtkPath.html) class may be what you want. However, I am trying to figure out how to achieve this in PyVista.

**Reply by tkoyama010** [reply]
The same may be done with the [tesselate](https://dev.pyvista.org/api/core/_autosummary/pyvista.StructuredGrid.tessellate.html#tessellate) filter, although the method may be different.
| before | after |
| ------------- | ------------- |
| ![pyvista-StructuredGrid-tessellate-1_00_00](https://user-images.githubusercontent.com/7513610/166202389-37dc9065-dd0b-4ed4-98a2-ead4b5f10b4f.png)  | ![pyvista-StructuredGrid-tessellate-1_01_00](https://user-images.githubusercontent.com/7513610/166202421-ab978445-90c6-4914-9aab-2aa840f006e4.png)  |

**Reply by stla** [reply]
Thanks! Looks nice. I gonna try.

**Reply by stla** [reply]
Your link is in the doc for **StructuredGrid** but I get:

> AttributeError: 'StructuredGrid' object has no attribute 'tessellate'

And the given example is for an **UnstructuredGrid**. Looks like this doc is not at the place it should be...

**Reply by stla** [reply]
Ah I also get:

> AttributeError: 'UnstructuredGrid' object has no attribute 'tessellate'

Looks like I have to update...

**Comment by stla** [comment]
Ok, works with pyvista 0.34.1.

This refines the mesh, indeed, but it is not as smooth as I expected.

```python
# -*- coding: utf-8 -*-
import numpy as np

import pyvista as pv


# A spherical curve
def scurve(t):
    alpha = np.pi / 2 - (np.pi / 2 - 0.44) * np.cos(3 * t)
    beta = t + 0.44 * np.sin(6 * t)
    return np.array([np.sin(alpha) * np.cos(beta), np.sin(alpha) * np.sin(beta), np.cos(alpha)])


# Hopf fiber
def hopf_fiber(p, phi):
    return np.array(
        [
            (1 + p[2]) * np.cos(phi),
            p[0] * np.sin(phi) - p[1] * np.cos(phi),
            p[0] * np.cos(phi) + p[1] * np.sin(phi),
            (1 + p[2]) * np.sin(phi),
        ]
    ) / np.sqrt(2 * (1 + p[2]))


# Stereographic projection
def stereo_proj(q):
    return q[0:3] / (1 - q[3])


# Parameterization of the Hopf torus
def hopf_torus(t, phi):
    return stereo_proj(hopf_fiber(scurve(t), phi))


# Create the mesh
angle_u = np.linspace(-np.pi, np.pi, 60)
angle_v = np.linspace(0, np.pi, 60)
u, v = np.meshgrid(angle_u, angle_v)
x, y, z = hopf_torus(u, v)
grid = pv.StructuredGrid(x, y, z).cast_to_unstructured_grid()

tessellated = grid.tessellate(max_n_subdivide=20)
tessellated.clear_data()  # cleans up plot
tessellated.plot(show_edges=True, line_width=1)
```

![HopfTessellated](https://user-images.githubusercontent.com/4466543/166210041-7ea7bfb0-099d-43a2-88e2-365f179e3ace.png)

**Reply by stla** [reply]
A little better with `smooth_shading=True`.

**Comment by stla** [comment]
I took a look at [the bezier library](https://bezier.readthedocs.io/en/stable/python/reference/bezier.triangle.html). I don't have the impression it could be helpful here. This smooths a triangle but it remains planar.

**Comment by stla** [comment]
There are amazing smoothers in VCG, or Meshlab. This one for example, but it changes the shape of the mesh:

![HTsmoothed](https://user-images.githubusercontent.com/4466543/168384534-40da4b16-667b-47b2-a4a3-f13538e1623d.png)
